### PR TITLE
Rename submodules under `cupy.testing` package

### PR DESCRIPTION
**Pull request by takagi**
Rel #3420. ~~Merge #3867 first.~~

This PR renames the submodules under `cupy.testing` to indicate that they are CuPy's private submodules and not intended to be used by users.

### CODE REVIEW

**Comment by takagi** [comment]
I'll resolve the conflicts.

**Comment by kmaehashi** [comment]
Could you check test failures?

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Should we keep `cupy.testing.attr` as a public submodule?

https://docs.cupy.dev/en/stable/contribution.html?highlight=attr#how-to-write-tests

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
> Should we keep cupy.testing.attr as a public submodule?

Let me hide `cupy.testing.attr` and other submodules here as @kmaehashi says hiding APIs that not described on the reference. Which APIs should be public is another problem.

**Comment by takagi** [comment]
Talked with @toslunar that we are fine to keep all the functions and decorators directly under `cupy.testing` this time.

And some future works:
- We may replace `parameterize` and `product` with those in PyTest
- We may remove `gpu` mark that was originally for Chainer. Only `multi_gpu(2)` is necessary now.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
PTAL, @toslunar!

File: install/build.py [hunk 1]
```diff
@@ -604,6 +604,7 @@ def check_cub_version(compiler, settings):
 
     _cub_version = int(out)
     settings['define_macros'].append(('CUPY_CUB_VERSION_CODE', _cub_version))
+    settings['define_macros'].append(('THRUST_IGNORE_CUB_VERSION_CHECK', 1))
```
Comment (toslunar) [inline install/build.py:607]: The change of the line seems irrelevant to the PR.

Comment (takagi) [reply install/build.py]: Removed! (maybe accidentally included...)

**Comment by takagi** [comment]
I'll fix the CI fails soon.

**Comment by takagi** [comment]
pfnCI, test this please.

**Review by toslunar — APPROVED** [review]

**Review by toslunar — APPROVED** [review]

**Comment by takagi** [comment]
Thanks!

**Comment by takagi** [comment]
Would you make a backport of this, @toslunar? (though, I'm not sure if it can be merged into v8 with less effort... If not, I will resolve the conflicts or manually make the backport and ask another for reviewing, so please run the backport script first!)
